### PR TITLE
feat: Automatic update of node version

### LINKED ISSUE #9864: Node.js auto-installed to .vaadin should be auto-upgraded
The Node.js version automatically downloaded and installed into `~/.vaadin` has recently been updated to latest LTS 14 (from 12). However, if version 12 was previously installed into `~/.vaadin` it will continue to take precedence until the dev deletes it from their `~/.vaadin` folder. ~Upgrading the global Node.js installation does not help either (because of #9863).~ Flow could automatically upgrade the Node.js version whenever it resorts to it (no global `npm` or forced via `require.home.node`).

### PR DESCRIPTION
**Pull request by caalador**
Enable automatic update of the node version
installed by the framework. Auto update is opt in,
but for a version that is older than what should
work it is updated automatically.

Closes #9864

### CODE REVIEW

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java [hunk 1]
```diff
@@ -116,14 +114,101 @@ public void installNode_NodeIsInstalledToTargetDirectory()
...
                 newTools.getNpmExecutable());
         npmVersionCommand.add("--version");
         FrontendVersion npm = FrontendUtils.getVersion("npm",
                 npmVersionCommand);
-        Assert.assertEquals("7.10.0", npm.getFullVersion());
+        Assert.assertEquals("7.20.3", npm.getFullVersion());
+    }
+
+    @Test
+    // @Ignore("Ignored to lessen PRs hitting the server too often")
```
Comment (mshabarov) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java:130]:
```suggestion
```

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java [hunk 2]
```diff
@@ -116,14 +114,101 @@ public void installNode_NodeIsInstalledToTargetDirectory()
...
+        nodeVersionCommand.add("--version");
+        node = FrontendUtils.getVersion("node", nodeVersionCommand);
+        Assert.assertEquals("Node version update failed.",
+                new FrontendVersion(FrontendTools.DEFAULT_NODE_VERSION)
+                        .getFullVersion(),
+                node.getFullVersion());
+    }
+
+    @Test
+    // @Ignore("Ignored to lessen PRs hitting the server too often")
```
Comment (mshabarov) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java:159]:
```suggestion
```

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java [hunk 3]
```diff
@@ -116,14 +114,101 @@ public void installNode_NodeIsInstalledToTargetDirectory()
...
+        nodeVersionCommand.add(nodeExecutable);
+        nodeVersionCommand.add("--version");
+        node = FrontendUtils.getVersion("node", nodeVersionCommand);
+        Assert.assertEquals(
+                "Node version updated even if it should not have been touched.",
+                "10.14.2", node.getFullVersion());
+    }
+
+    @Test
+    // @Ignore("Ignored to lessen PRs hitting the server too often")
```
Comment (mshabarov) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java:186]:
```suggestion
```

**Review by mshabarov — APPROVED** [review]
Looks good to me. Too old Node versions will be always updated.  Supported but older than default versions are optionally auto-installed, so this allows to swap between V14 and V22+ projects with no re-installations (unless the machine has really old Node).

**Review by mshabarov — APPROVED** [review]

**Comment by Artur-** [comment]
Why did you break public API here? Doesn't seem necessary at all

**Comment by caalador** [comment]
FrontendTools? Because it was impossible to manage all the overrides, they didn't work together anymore and its an internal public class.

**Pull request merged**